docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,387 @@
+  attributes can be lazy as described later in the tutorial.
+* :meth:`apply`: you need to implement a method that actually
+  implements the operation of the brick, taking as arguments the inputs
+  of the brick and returning its outputs. It can have any name and for simple
+  bricks is often named ``apply``. You should decorate it with the
+  :func:`.application` decorator, as explained in the next section. If you
+  design a recurrent brick, you should instead decorate it with the
+  :func:`.recurrent` decorator as explained in the
+  :doc:`tutorial about rnns</rnn>`.
+* :meth:`.Brick._allocate`: you should implement this method to allocate the
+  shared variables (often representing parameters) of the brick. In blocks,
+  by convention, the built-in bricks allocate their shared variables with nan
+  values and we recommend you to do the same.
+* :meth:`.Brick._initialize`: you should implement this method to initialize
+  the shared variables of your brick. This method is called after the
+  allocation.
+* :meth:`.Brick._push_allocation_config`: you should consider overwriting
+  this method if you want to change configuration of the children bricks
+  before they allocate their parameters.
+* :meth:`.Brick._push_initialization_config`: you should consider
+  overwriting this method if you want to change the initialization schemes of
+  the children before they get initialized.
+  If the children bricks need to be initialized with the same scheme, then you
+  should inherit your brick from :class:`.Initializable`, which
+  automatically push the initialization schemes of your brick (provided as
+  arguments ``weights_init`` and ``biases_init`` of the constructor) to the
+  children bricks.
+* :meth:`.Brick.get_dim`: implementing this function is useful if you want
+  to provide a simple way to get the dimensions of the inputs and outputs of
+  the brick.
+
+If you want to inherit from a specific brick, check its docstring to
+identify the particular methods to overwrite and the attributes to define.
+
+Application methods
+~~~~~~~~~~~~~~~~~~~
+
+The :meth:`apply` method listed above is probably the most
+important method of your brick because it is the one that actually takes
+theano tensors as inputs, process them and return output tensors. You should
+decorate it with the :func:`.application` decorator, which names variables
+and register auxiliary variables of the operation you implement.
+It is used as follows:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application(inputs=['input1', 'input2'], outputs=['output'])
+    ...     def apply(self, input1, input2)
+    ...         ...
+    ...         return something
+
+In the case above, it will automatically label the theano tensor variable
+``input1`` to ``Foo_apply_input1``, idem for ``input2`` and the output of the
+method.
+
+Under the hood, the ``@application`` decorator creates an object of class
```

Comment:
Also, a small code fragment would be helpful (which queries `Foo.apply` and shows what kind of object it is).